pkg/kubectl/cmd/edit.go

```diff
@@ -0,0 +1,409 @@
+			return preservedFile(err, file, out)
+		} else {
+			preconditions = append(preconditions, jsonmerge.RequireKeyUnchanged("apiVersion"))
+			preconditions = append(preconditions, jsonmerge.RequireKeyUnchanged("kind"))
+			preconditions = append(preconditions, jsonmerge.RequireMetadataKeyUnchanged("name"))
+			results.version = defaultVersion
+		}
+
+		//TODO: need a more strict check for lists
+		if hold, msg := TestPreconditionsHold(patch, preconditions); !hold {
+			fmt.Fprintf(out, "error: %s", msg)
+			return preservedFile(nil, file, out)
+		}
+
+		err = visitor.Visit(func(info *resource.Info, err error) error {
+			patch, err = strategicpatch.StrategicMergePatch(originalJS, patch, obj)
+//			patched, err := resource.NewHelper(info.Client, info.Mapping).Patch(info.Namespace, info.Name, api.StrategicMergePatchType, patch)
+			patched, err := resource.NewHelper(info.Client, info.Mapping).Patch(info.Namespace, info.Name, api.MergePatchType, patch)
+			if err != nil {
+				fmt.Fprintln(out, results.AddError(err, info))
+				return nil
+			}
+			info.Refresh(patched, true)
+			cmdutil.PrintSuccess(mapper, false, out, info.Mapping.Resource, info.Name, "edited")
+			return nil
+		})
+		if err != nil {
+			return preservedFile(err, file, out)
+		}
+
+		if results.retryable > 0 {
+			fmt.Fprintf(out, "You can run `kubectl replace -f %s` to try this update again.\n", file)
+			return errExit
+		}
+		if results.conflict > 0 {
+			fmt.Fprintf(out, "You must update your local resource version and run `kubectl replace -f %s` to overwrite the remote changes.\n", file)
+			return errExit
+		}
+		if len(results.edit) == 0 {
+			if results.notfound == 0 {
+				os.Remove(file)
+			} else {
+				fmt.Fprintf(out, "The edits you made on deleted resources have been saved to %q\n", file)
+			}
+			return nil
+		}
+
+		// loop again and edit the remaining items
+		infos = results.edit
+	}
+	return nil
+}
+
+// TestPreconditions test if preconditions hold given the edit
+func TestPreconditionsHold(edit []byte, preconditions []jsonmerge.PreconditionFunc) (bool, string) {
+	diff := make(map[string]interface{})
+	if err := json.Unmarshal(edit, &diff); err != nil {
+		return false, err.Error()
+	}
+	for _, fn := range preconditions {
+		if hold, msg := fn(diff); !hold {
+			return false, msg
+		}
+	}
+	return true, ""
+}
+
+// print json file (such as patch file) content for debugging
+func printJson(out io.Writer, file []byte) error {
+	diff := make(map[string]interface{})
+	if err := json.Unmarshal(file, &diff); err != nil {
+		return err
+	}
+	fmt.Fprintf(out, "%v\n", diff)
+	return nil
+}
+
+// editReason preserves a message about the reason this file must be edited again
+type editReason struct {
+	head  string
+	other []string
+}
+
+// editHeader includes a list of reasons the edit must be retried
+type editHeader struct {
+	reasons []editReason
+}
+
+// WriteTo outputs the current header information into a stream
```

Comment:
I think this should be `writeTo` unless there's a good reason to make it public (and if so, should probably be moved to util/editor/editor.go).